client/boot/index.js

```diff
@@ -369,6 +370,30 @@ function reduxStoreReady( reduxStore ) {
 		require( 'lib/rubberband-scroll-disable' )( document.body );
 	}
 
+	// 404 handler.
+	page( '*', function( context, next ) {
+		let Page404 = require( 'layout/404.jsx' );
+
+		if ( ! config.isEnabled( 'code-splitting' ) ) {
+			// If code-splitting is not enabled and we got to this route handler, no section route was matched.
+			Page404.show();
```

Comment:
Good question. To be honest, I took an example of `client/layout/error.jsx`. The main reason why I liked this approach was because you could render the 404 component to string with different static method in the future, thus making the component reusable on server-side (if we ever needed 404 there). Encapsulate the 404 behavior in one `404.jsx` file, which contains the component and some static methods to render it.